server/asset/eth/contracts/ETHSwapV0.sol

```diff
@@ -0,0 +1,84 @@
+// SPDX-License-Identifier: BlueOak-1.0.0
+pragma solidity >=0.7.0 <0.9.0;
+contract ETHSwap {
+    enum State { Empty, Filled, Redeemed, Refunded }
+
+    struct Swap {
+        uint initBlockNumber;
+        uint refundBlockTimestamp;
+        bytes32 secretHash;
+        bytes32 secret;
+        address initiator;
+        address participant;
+        uint256 value;
+        State state;
+    }
+
+    mapping(bytes32 => Swap) public swaps;
+
+    constructor() {}
+
+    modifier isRefundable(bytes32 secretHash, address refunder) {
+        require(swaps[secretHash].state == State.Filled);
+        require(swaps[secretHash].initiator == refunder);
+        uint refundBlockTimestamp = swaps[secretHash].refundBlockTimestamp;
+        require(block.timestamp >= refundBlockTimestamp);
+        _;
+    }
+
+    modifier isRedeemable(bytes32 secretHash, bytes32 secret, address redeemer) {
+        require(swaps[secretHash].state == State.Filled);
+        require(swaps[secretHash].participant == redeemer);
+        require(sha256(abi.encodePacked(secret)) == secretHash);
+        _;
+    }
+
+    modifier isNotInitiated(bytes32 secretHash) {
+        require(swaps[secretHash].state == State.Empty);
+        _;
+    }
+
+    modifier hasNoNilValues(uint refundTime) {
+        require(msg.value > 0);
+        require(refundTime > 0);
+        _;
+    }
+
+    function swap(bytes32 secretHash)
+        public view returns(Swap memory)
+    {
+        return swaps[secretHash];
+    }
+
+    function initiate(uint refundTimestamp, bytes32 secretHash, address participant)
```

Comment:
It might save some gas fees if there’s a batch version of this function that can be used in case there are multiple matches in the same epoch.

Reply:
I think that's true. @chappjc want to add that functionality to the contract? Can also do it in V1 if we want to keep things as simple as possible for now.

Reply:
I didn't bring it up in this context, but it may be worth investigating.  It's not clear to me that gas use is reduced by batching in the same way that tx size and thus fees in utxo assets is reduced, but worth considering.

Reply:
Maybe we defer this idea though unless it's really simple and we clearly characterize the gas consumption changes for various cases.

Reply:
Actually, would prefer to keep things lean and simple as possible for the first interation. Just needs to work and be secure.